File: bench/latency/latency.go
// Latency angle, Go: the working set is garbage each round; the GC must collect it.
// Reports its own GC count + total pause time (stderr) — the pause cost hier/C don't pay.
package main
import ("fmt"; "os"; "runtime")
func main() {
    const iters, k = 2000, 100000
    acc := 0
    for it := 0; it < iters; it++ {
        xs := make([]int, 0)
        for j := 0; j < k; j++ { xs = append(xs, (it+j)%997) }
        s := 0
        for _, v := range xs { s += v }
        acc = (acc + s) % 1000000007
    }
    var ms runtime.MemStats
    runtime.ReadMemStats(&ms)
    fmt.Fprintf(os.Stderr, "gc=%d pause_us=%d\n", ms.NumGC, ms.PauseTotalNs/1000)
    fmt.Println(acc)
}

File: bench/latency/latency.c
/* Latency angle, C: malloc/free the working set each round -> no GC, predictable.
 * Same checksum as the hier port. */
#include <stdio.h>
#include <stdlib.h>
int main(void) {
    long iters = 2000, k = 100000, acc = 0;
    for (long it = 0; it < iters; it++) {
        long *xs = malloc((size_t)k * sizeof(long));
        for (long j = 0; j < k; j++) xs[j] = (it + j) % 997;
        long s = 0;
        for (long m = 0; m < k; m++) s += xs[m];
        acc = (acc + s) % 1000000007L;
        free(xs);
    }
    printf("%ld\n", acc);
    return 0;
}
